pkg/services/store/resolver/service.go

```diff
@@ -0,0 +1,122 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+// for testing
+var getNow = func() time.Time { return time.Now() }
+
+type ResolutionInfo struct {
+	OK        bool      `json:"ok"`
+	Key       string    `json:"key,omitempty"`  // GRN? UID?
+	Warning   string    `json:"kind,omitempty"` // old syntax?  (name>uid) references a renamed object?
+	Timestamp time.Time `json:"timestamp,omitempty"`
+}
+
+type ObjectReferenceResolver interface {
+	Resolve(ctx context.Context, ref *models.ObjectExternalReference) (ResolutionInfo, error)
+}
+
+func ProvideObjectReferenceResolver(ds datasources.DataSourceService, pluginRegistry registry.Service) ObjectReferenceResolver {
+	return &standardReferenceResolver{
+		pluginRegistry: pluginRegistry,
+		ds: dsCache{
+			ds:             ds,
+			pluginRegistry: pluginRegistry,
+		},
+	}
+}
+
+type standardReferenceResolver struct {
+	pluginRegistry registry.Service
+	ds             dsCache
+}
+
+func (r *standardReferenceResolver) Resolve(ctx context.Context, ref *models.ObjectExternalReference) (ResolutionInfo, error) {
+	if ref == nil {
+		return ResolutionInfo{
+			OK:        false,
+			Timestamp: getNow(),
+			Warning:   "invalid reference (nil)",
```

Comment:
should this be just an `error` rather than a warning?

Reply:
yup -- updated and added test